### PR TITLE
use single string dict and eliminate the integer key field

### PR DESCRIPTION
**Pull request by walterddr**
follow up on #8890. 

We currently have multiple Map of Map of int -> String mapping for data table. 
since we encode the integer key field in fix-sized buffer. it can be simplified into just a `String[]`.

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableBuilder.java [hunk 1]
```diff
@@ -99,29 +97,32 @@ public class DataTableBuilder {
       new DataOutputStream(_variableSizeDataByteArrayOutputStream);
 
   private int _numRows;
+  private int _numDictionaryEntries;
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableBuilder.java:100]: For V4, let's directly update a single `Map<String, Integer>` dictionary and a String[] reverse dictionary

[the remark this answers is not in the document]
Comment (walterddr) [reply pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableBuilder.java]: yeah it has some inefficiency but the approach is correct. let me refactor.

File: pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableFactory.java [hunk 2]
```diff
@@ -20,22 +20,58 @@
...
+
   private DataTableFactory() {
   }
 
+  public static int getCurrentDataTableVersion() {
+    return _version;
+  }
+
+  public static void setCurrentDataTableVersion(int version) {
+    LOGGER.warn("Setting current DataTable version to: " + version);
```
Comment (jackjlli) [inline pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableFactory.java]: Why warn level here?

File: pinot-core/src/main/java/org/apache/pinot/core/common/datablock/BaseDataBlock.java [hunk 3]
```diff
@@ -387,24 +375,14 @@ protected byte[] serializeDictionaryMap()
...
-      for (int j = 0; j < dictionarySize; j++) {
-        int key = buffer.getInt();
-        String value = DataTableUtils.decodeString(buffer);
-        dictionary.put(key, value);
-      }
-      dictionaryMap.put(column, dictionary);
+    int dictionarySize = buffer.getInt();
+    String[] stringDictionary = new String[dictionarySize];
+    for (int i = 0; i < dictionarySize; i++) {
+      stringDictionary[i] = (DataTableUtils.decodeString(buffer));
```
Comment (jackjlli) [inline pinot-core/src/main/java/org/apache/pinot/core/common/datablock/BaseDataBlock.java:383]: Remove outer `()`?

File: pinot-core/src/main/java/org/apache/pinot/core/common/DataTableFactory.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+
+public class DataTableFactory {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DataTableFactory.class);
+  public static final int VERSION_2 = 2;
+  public static final int VERSION_3 = 3;
+  public static final int VERSION_4 = 4;
+  public static final int DEFAULT_VERSION = VERSION_3;
+
+  private static int _version = DataTableFactory.DEFAULT_VERSION;
```
Comment (jackjlli) [inline pinot-core/src/main/java/org/apache/pinot/core/common/DataTableFactory.java:41]: put `private static` all together. Also make the constant name upper cases?

File: pinot-core/src/main/java/org/apache/pinot/core/common/datablock/BaseDataBlock.java [hunk 5]
```diff
@@ -84,7 +84,7 @@ public abstract class BaseDataBlock implements DataTable {
   protected int _numRows;
   protected int _numColumns;
   protected DataSchema _dataSchema;
-  protected Map<String, Map<Integer, String>> _dictionaryMap;
+  protected String[] _stringDictionary;
```
Comment (jackjlli) [inline pinot-core/src/main/java/org/apache/pinot/core/common/datablock/BaseDataBlock.java:87]: `_stringDictionaries`?

File: pinot-core/src/main/java/org/apache/pinot/core/common/DataTableFactory.java (hunk 4 above)
Comment (walterddr) [reply pinot-core/src/main/java/org/apache/pinot/core/common/DataTableFactory.java:41]: this field is not final. thus individually separated

**Review by Jackie-Jiang — APPROVED** [review]

**Pull request merged**